signer/src/network/libp2p/network.rs

```diff
@@ -206,4 +209,239 @@ mod tests {
+            current_signer_set3.add_signer(PublicKey::from_private_key(&key));
+        }
+
+        let mut swarm1 = SignerSwarmBuilder::new(&key1, true)
+            .add_listen_endpoint("/ip4/0.0.0.0/tcp/25001".parse().unwrap())
+            .add_seed_addr("/ip4/0.0.0.0/tcp/25002".parse().unwrap())
+            .build()
+            .expect("Failed to build swarm 1");
+
+        let mut swarm2 = SignerSwarmBuilder::new(&key2, true)
+            .add_listen_endpoint("/ip4/0.0.0.0/tcp/25002".parse().unwrap())
+            .add_seed_addr("/ip4/0.0.0.0/tcp/25001".parse().unwrap())
+            .add_seed_addr("/ip4/0.0.0.0/tcp/25003".parse().unwrap())
+            .build()
+            .expect("Failed to build swarm 2");
+
+        let mut swarm3 = SignerSwarmBuilder::new(&key3, true)
+            .add_listen_endpoint("/ip4/0.0.0.0/tcp/25003".parse().unwrap())
+            .add_seed_addr("/ip4/0.0.0.0/tcp/25002".parse().unwrap())
+            .build()
+            .expect("Failed to build swarm 3");
+
+        let mut network1 = P2PNetwork::new(&context1);
+        let _network2 = P2PNetwork::new(&context2);
+        let mut network3 = P2PNetwork::new(&context3);
+
+        // Start three swarms.
+        tokio::spawn(async move {
+            swarm1.start(&context1).await.unwrap();
+        });
+        tokio::spawn(async move {
+            swarm2.start(&context2).await.unwrap();
+        });
+        tokio::spawn(async move {
+            swarm3.start(&context3).await.unwrap();
+        });
+
+        // The swarms are discovering themselves via mDNS, so we need to give
+        // them a bit of time to connect.
+        tokio::time::sleep(Duration::from_secs(1)).await;
+
+        let number_of_messages = 10;
+        let mut signed_messages: Vec<Msg> = std::iter::repeat_with(|| Msg::random(&mut rng))
+            .take(number_of_messages)
+            .collect();
+        signed_messages.sort_by_cached_key(|x| x.inner.bitcoin_chain_tip);
+
+        let (broadcast_signer_msg, rx) = tokio::sync::broadcast::channel(20);
+        let signer_msg_stream = BroadcastStream::new(rx);
+
+        let broadcast_messages = signed_messages.clone();
+        tokio::spawn(async move {
+            for msg in broadcast_messages {
+                network3.broadcast(msg).await.expect("Failed to broadcast");
+            }
+        });
+
+        tokio::spawn(async move {
+            loop {
+                let message = network1.receive().await.expect("Failed to receive message");
+                broadcast_signer_msg.send(message).unwrap();
+            }
+        });
+
+        // We wait for our first message, but not for too long, life is short.
+        let received_messages = signer_msg_stream
+            .take(1)
+            .take_until(tokio::time::sleep(Duration::from_secs(4)))
+            .collect::<Vec<_>>()
+            .await;
+
+        assert!(received_messages.is_empty());
```

Comment:
nit: while I'm pretty confident about the small change, can we add in here (even in a different PR) a check that signer 2 is getting signer 3 messages, and that if signer 2 sends a message signer 1 will receive it? To ensure that the test is not passing just because signer 1 is not connected at all

Reply:
Good idea, I think I'll do this here.